website/pages/en/cookbook/substreams-powered-subgraphs.mdx

```diff
@@ -0,0 +1,242 @@
+                .uncle_rewards
+                .unwrap()
+                .deserialize()
+                .to_field("blockUncleReward"),
+            block_delta
+                .burned
+                .unwrap()
+                .deserialize()
+                .to_field("blockBurnedFees"),
+        ],
+    }];
+    Ok(EntityChanges { entity_changes })
+}
+```
+
+An `EntityChange` must specify the entity type, ID (which can be a String, or Bytes, in line with subgraph entity IDs), operation, and fields.
+
+> The ID and fields must be compatible with the `schema.graphql` definition on the corresponding subgraph for the entity type.
+
+The operation can be "Create", "Update" or "Delete".
+
+## subgraph.yaml
+
+Substreams-powered subgraphs introduce a new `kind` of data source, "substreams". Such subgraphs can only have one data source. This data source must specify the Substreams network, the Substreams package as a relative file location, and the module within that Substreams package which produces subgraph-compatible entity changes (in this case `map_entity_changes`, from the Substreams package above). The mapping is specified, but simply identifies the mapping kind ("substreams/graph-entities") and the apiVersion.
+
+```yaml
+specVersion: 0.0.4
+description: Ethereum supply subgraph powered by substreams
+repository: https://github.com/messari/substreams/eth-supply
+schema:
+  file: ./schema.graphql
+
+dataSources:
+  - kind: substreams
+    name: eth_supply
+    network: mainnet
+    source:
+      package:
+        moduleName: map_entity_changes
+        file: eth-supply-v0.1.0.spkg
+    mapping:
+      kind: substreams/graph-entities
+      apiVersion: 0.0.5
+```
+
+## schema.graphql
+
+The `subgraph.yaml` also references a schema file. The requirements for this file are unchanged, but the entities specified must be compatible with the entity changes produced by the Substreams module referenced in the `subgraph.yaml`.
+
+```graphql
+type Supply @entity(immutable: true) {
+  id: Bytes!
+
+  blockHash: Bytes!
+  blockNumber: BigInt!
+  timestamp: BigInt!
+
+  " Total ETH in existence at this block. All initial supply and rewards added up, minus burned fees. "
+  currentSupply: BigInt!
+
+  " ETH created at genesis. "
+  genesisSupply: BigInt!
+
+  " ETH created as block mining rewards (up until the merge). "
+  cumulativeMiningRewards: BigInt!
+
+  " ETH created as part of including uncle blocks (up until the merge). "
+  cumulativeUncleRewards: BigInt!
+
+  " ETH burned because of base fee burn after eip1559. "
+  cumulativeBurnedFees: BigInt!
+
+  " ETH created in this block as mining rewards. "
+  blockMiningReward: BigInt!
+
+  " ETH created in this block as uncle rewards. "
+  blockUncleReward: BigInt!
+
+  " Eth burned in this block via eip1559. "
+  blockBurnedFees: BigInt!
+}
+```
+
+Given the above, subgraph developers can use Graph CLI >=0.51.0 to build and deploy this Substreams-powered subgraph.
+
+```bash
+graph build
```

Comment:
Does `graph build` detect that it is working with substreams? Compiling & packing it into the `spkg`?

I've been doing `cargo build` + `substreams pack` + `graph deploy` all this time 😅

Reply:
ah - good catch, `graph build` actually isn't necessary in this step (you have been doing the right thing).

Reply:
But with `substreams-js`... we actually could build the `spkg` as part of the Graph CLI in the future (?)

Reply:
If that doesn't work, I'd add a note to remind people of building and packing for their changes to be reflected in the subgraph they deploy (or to work at all the first time).

> But with substreams-js... we actually could build the spkg as part of the Graph CLI in the future (?)

This could be good :)